### PR TITLE
Fix Discv5 in Runtime 

### PR DESCRIPTION
**Pull request by nisdas**
Part of #3147

- Make only the config value `port` be used for discovery instead of `UDPPort`
- Make the private key a port of the main service struct, so there is only one version of the private key.
- Add disconnection handler to not dial previously disconnected peers again

### CODE REVIEW

File: beacon-chain/p2p/service.go [hunk 1]
```diff
@@ -179,20 +187,16 @@ func (s *Service) Disconnect(pid peer.ID) error {
...
 		select {
 		case <-ticker.C:
-			nodes := s.dv5Listener.Lookup(nodeID)
-			multiAddresses := convertToMultiAddr(nodes)
+			num := s.dv5Listener.ReadRandomNodes(nodes)
+			multiAddresses := convertToMultiAddr(nodes[:num])
 			s.connectWithAllPeers(multiAddresses)
 			// store furthest node as the next to lookup
-			nodeID = nodes[len(nodes)-1].ID
+			//nodeID = nodes[len(nodes)-1].ID
```
Comment (prestonvanloon) [inline beacon-chain/p2p/service.go]: Remove?

File: beacon-chain/p2p/discovery.go [hunk 2]
```diff
@@ -60,32 +60,38 @@ func startDiscoveryV5(addr net.IP, privKey *ecdsa.PrivateKey, cfg *Config) (*dis
...
 }
 
+func convertToSingleMultiAddr(node *discv5.Node) (ma.Multiaddr, error) {
+	ip4 := node.IP.To4()
+	if ip4 == nil {
+		return nil, errors.New("node doesn't have an ip4 address")
+	}
+	pubkey, err := node.ID.Pubkey()
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not get pubkey from node ID")
```
Comment (terencechain) [inline beacon-chain/p2p/discovery.go]:
```suggestion
		return nil, errors.Wrap(err, "could not get pubkey from node ID")
```

File: beacon-chain/p2p/discovery.go [hunk 3]
```diff
@@ -60,32 +60,38 @@ func startDiscoveryV5(addr net.IP, privKey *ecdsa.PrivateKey, cfg *Config) (*dis
...
+	}
+	assertedKey := convertToInterfacePubkey(pubkey)
+	id, err := peer.IDFromPublicKey(assertedKey)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not get peer id")
```
Comment (terencechain) [inline beacon-chain/p2p/discovery.go]:
```suggestion
		return nil, errors.Wrap(err, "could not get peer id")
```

File: beacon-chain/p2p/discovery.go [hunk 4]
```diff
@@ -60,32 +60,38 @@ func startDiscoveryV5(addr net.IP, privKey *ecdsa.PrivateKey, cfg *Config) (*dis
...
+	}
+	multiAddrString := fmt.Sprintf("/ip4/%s/tcp/%d/p2p/%s", ip4.String(), node.TCP, id)
+	multiAddr, err := ma.NewMultiaddr(multiAddrString)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not get multiaddr")
```
Comment (terencechain) [inline beacon-chain/p2p/discovery.go]:
```suggestion
		return nil, errors.Wrap(err, "could not get multiaddr")
```

File: beacon-chain/p2p/service.go (hunk 1 above)
Comment (nisdas) [reply beacon-chain/p2p/service.go]: Done

**Pull request merged**